lib/voight_kampff/test.rb

```diff
@@ -1,6 +1,77 @@
+# frozen_string_literal: true
+
+require 'net/http'
+require 'json'
+
 module VoightKampff
+  # Test User-Agent against Voight-Kampff
   class Test
-    CRAWLERS_FILENAME = 'crawler-user-agents.json'
+    DEFAULT_URL =
+      'https://raw.githubusercontent.com/monperrus/crawler-user-agents/master/crawler-user-agents.json'
+    CRAWLERS_FILENAME = File.basename(DEFAULT_URL).freeze
+
+    class << self
+      def cache_path
+        @cache_path ||= "#{Dir.pwd}/tmp/#{CRAWLERS_FILENAME}"
+      end
+
+      def cache_path=(value)
+        @cache_path = File.expand_path value
+      end
+
+      def max_age
+        @max_age ||= 24 * 3600
+      end
+
+      attr_writer :max_age
+
+      def crawlers
+        return @crawlers if @crawlers && fresh?(@last_updated_at)
+        @last_updated_at = Time.now
+        @crawler_regexp = nil
+        @crawlers = JSON.parse(load_crawlers)
+      end
+
+      def crawler_regexp
+        @crawler_regexp ||= begin
+          # NOTE: This is admittedly a bit convoluted
+          # but the performance gains make it worthwhile
+          crawler_patterns =
+            crawlers.map.with_index do |crawler, index|
+              "(?<match#{index}>#{crawler['pattern']})"
+            end.join('|')
+          crawler_patterns = "(#{crawler_patterns})"
+          Regexp.new(crawler_patterns, Regexp::IGNORECASE)
+        end
+      end
+
+      def load_crawlers
+        load_local_crawlers || load_remote_crawlers
+      end
+
+      private
+
+      def fresh?(timestamp)
+        return false unless timestamp
+        timestamp >= Time.now - max_age
+      end
+
+      def load_remote_crawlers(uri = DEFAULT_URL)
+        crawlers = Net::HTTP.get(URI(uri)).force_encoding('UTF-8')
```

Comment:
Do you think this should potentially fail gracefully if say the URL returns a non successful response code or any kind of network error? Potentially just `Kernel.warn "VoightKampff: Unable to fetch crawlers from #{uri}"`

I can see a scenario where the repo gets pulled and a bunch of sites go down.

Reply:
Wow, thank you. I added tests and code for these situations.

Also, there is `warn` with error inspection: it can be `SocketError`, or `Errno::ENETUNREACH`, or something else.

And the code will "cache" the empty Array in these cases.

Also, I didn't add a code for redirects: `Net::HTTP` can't deal with them easily. It's more comfortable to use some external gem, like [`httpx`](https://gitlab.com/honeyryderchuck/httpx), for this. But I hope that there will be not redirects.

Reply:
Yes, if we did this at all, I'd want to be really paranoid about it. It should fail very gracefully and any sort of issues with this whole process should log an error, but never take down the whole site. And I'd prefer to avoid a situation where we need to give false positives too. If there are problems getting the file, I suppose you might need to treat everything as not a bot until the file did download. But that seems bad too.

At the very least, I think the gem should start off using it's local file thats packaged with the gem. Then only if a new version is downloaded successfully and the JSON is parsed successfully and then some sort of sanity check is done to make sure we have a real list of bots (even if that's just checking the number of bots in the file). Then and only then, should we start to use that new list.